File: day06/solver.py
import sys
from collections import Counter

sys.path.insert(0, "../utils/py")
import utils


measure_time = utils.stopwatch()


@measure_time
def parse(raw_data):
    return [int(x) for x in raw_data.split(",")]


def life(counter, days):
    for _ in range(days):
        # print(f"day {_}: {counter}")
        if counter[0]:
            counter[7] += counter[0]
            counter[9] += counter[0]
            counter[0] = 0

        counter = Counter({
            k-1: v for k, v in counter.items() if k != 0
        })
    return counter


# PART 1
@measure_time
def solve1(data):
    counter = Counter(data)
    counter = life(counter, 80)
    return sum(counter.values())


# PART 2
@measure_time
def solve2(data):
    counter = Counter(data)
    counter = life(counter, 256)
    return sum(counter.values())


if __name__ == "__main__":
    data = parse(open("input.txt").read().strip())
    print("Part 1: {}".format(solve1(data)))
    print("Part 2: {}".format(solve2(data)))

    print("\nTime taken:")
    for func, time in measure_time.times:
        print(f"{func:8}{time}s")
    print("----------------")
    print("total   {}s".format(sum(t for _, t in measure_time.times)))

File: day06/solver.hs
import Data.List
import Utils

getInput :: IO String
getInput = do
    contents <- readFile "input.txt"
    return contents


type Parsed = [Int]
type Sol1 = Int
type Sol2 = Int



parse :: String -> Parsed
parse = Utils.parseInts


step :: [Int] -> [Int]
step (n0:n1:n2:n3:n4:n5:n6:n7:n8:[]) = n1:n2:n3:n4:n5:n6:(n7+n0):n8:n0:[]

life :: Int -> [Int] -> [Int]
life 0 fish = fish
life d fish = life (d-1) (step fish)


fish :: Parsed -> [Int]
fish parsed = [length . filter (==i) $ parsed | i <- [0..8]]


solve1 :: Parsed -> Sol1
solve1 = sum . life 80 . fish


solve2 :: Parsed -> Sol2
solve2 = sum . life 256 . fish


testdata = "3,4,3,1,2"
testresult1 = 5934
testresult2 = 26984457539
test1 = test (solve1 . parse) testresult1 testdata
test2 = test (solve2 . parse) testresult2 testdata


main = do
    parsed <- parse <$> getInput
    putStr "Part 1: "
    print . solve1 $ parsed
    putStr "Part 2: "
    print . solve2 $ parsed
